### ISSUE TITLE
[feature request] Factory default for extension attributes

### ISSUE THREAD

**Issue by Evpok** [explosion/spaCy#3036]
## Feature description
As usual with Python, mutable defaults are a big no-no for extension attributes, since they are shared by all instances, which leads to subtle and ~~quick to anger~~ hard to root out bugs (see e.g. #2581).
The documentation mentions that pitfall, but doesn't offer a convenient solution: if I want to keep a static list of interesting spans in my document in a `Doc._.interesting`, it is not clear where the getter and setter that I am supposed to use for this property should store the state. (From what I understand, it should probably be somewhere in `Doc.user_data`, but I have not found a lot of doc on that either)

I propose a `factory` argument to `set_extension` that would be called the first time that the value for the corresponding extension property is retrieved for a given instance (as `collections.defaultdict` does), so one could just write

```python
spacy.tokens.Doc.set_extension('interesting', factory=list)
```

**Comment by ines** [comment]
Thanks, this is a nice idea!

> From what I understand, it should probably be somewhere in `Doc.user_data`, but I have not found a lot of doc on that either

`Doc.user_data` is a dict that you can write to. So you can do something like this:

```python
doc = nlp("This is a doc")
doc.user_data["interesting"] = []
doc.user_data["interesting"].append("cool stuff")
```

**Comment by ines** [comment]
Resolved in #3389! Mutable defaults now work via the regular `default` keyword argument 🙂
